### PR TITLE
DT-235 bed move api

### PR DESCRIPTION
**Pull request by mikehalmamoj**
Define the API for the new bed move API (which could also be called a cell move or assigning a new living unit).

### CODE REVIEW

File: src/test/java/net/syscon/elite/api/resource/impl/OffenderMovementResourceApiTest_moveToCell.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.*;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static org.springframework.http.HttpMethod.PUT;
+import static org.springframework.http.HttpStatus.*;
+
+// TODO DT-235 These tests are here to help define the API, but currently using canned data.  Once MovementUpdateService is implemented, replace below with real data.
+public class OffenderMovementResourceApiTest_moveToCell extends ResourceTest {
```
Comment (mikehalmamoj) [inline src/test/java/net/syscon/elite/api/resource/impl/OffenderMovementResourceApiTest_moveToCell.java]: I tried to implement this as an @WebMvcTest but that involved major surgery on the configuration to isolate the web components and I deemed it too big a change for a test. :(

File: src/main/java/net/syscon/elite/api/resource/OffenderMovementsResource.java [hunk 2]
```diff
@@ -5,20 +5,13 @@
 import io.swagger.annotations.ApiParam;
 import io.swagger.annotations.ApiResponse;
 import io.swagger.annotations.ApiResponses;
-import net.syscon.elite.api.model.CourtHearing;
-import net.syscon.elite.api.model.CourtHearings;
-import net.syscon.elite.api.model.ErrorResponse;
-import net.syscon.elite.api.model.PrisonToCourtHearing;
+import net.syscon.elite.api.model.*;
```
Comment (matt-crowson-moj) [inline src/main/java/net/syscon/elite/api/resource/OffenderMovementsResource.java]: I think agreement for team going forwards is no wildcard imports.

File: src/main/java/net/syscon/elite/api/resource/OffenderMovementsResource.java [hunk 3]
```diff
@@ -47,4 +40,17 @@ CourtHearings getCourtHearings(
...
+    @PutMapping("/{bookingId}/living-unit/{livingUnitId}")
+    @ApiResponses(value = {
+            @ApiResponse(code = 200, message = "OK", response = OffenderBooking.class),
+            @ApiResponse(code = 400, message = "Invalid request.", response = ErrorResponse.class, responseContainer = "List"),
+            @ApiResponse(code = 404, message = "Requested resource not found.", response = ErrorResponse.class, responseContainer = "List"),
+            @ApiResponse(code = 500, message = "Unrecoverable error occurred whilst processing request.", response = ErrorResponse.class, responseContainer = "List")})
+    OffenderBooking moveToCell(
+            @ApiParam(value = "The offender booking id", example = "1200866") @PathVariable("bookingId") Long bookingId,
+            @ApiParam(value = "The cell location the offender has been moved to", example = "123123") @PathVariable("livingUnitId") Long livingUnitId,
+            @ApiParam(value = "The reason code for the move", example = "ADM", required = true) @RequestParam("reasonCode") String reasonCode, // TODO DT-235 Find out hwo the reason code is validated and mention it in the API docs
```
Comment (Mjwillis) [inline src/main/java/net/syscon/elite/api/resource/OffenderMovementsResource.java]: typo

File: src/main/java/net/syscon/elite/service/MovementUpdateService.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import org.springframework.transaction.annotation.Transactional;
+
+import java.time.LocalDateTime;
+
+@Service
+@Transactional
+public class MovementUpdateService {
+
+    // TODO DT-235 Implement this service - currently only exists to mock in the API tests
+    public OffenderBooking moveToCell(final Long bookingId, final Long livingUnitId, final String reasonCode, final LocalDateTime dateTime) {
```
Comment (Mjwillis) [inline src/main/java/net/syscon/elite/service/MovementUpdateService.java]: Will need `@VerifiyBookingAccess`

File: src/main/java/net/syscon/elite/api/resource/impl/OffenderMovementsResourceImpl.java [hunk 5]
```diff
@@ -33,4 +44,42 @@ public CourtHearing prisonToCourt(final Long bookingId, final Long courtCaseId,
...
+    @Override
+    public OffenderBooking moveToCell(final Long bookingId, final Long livingUnitId, final String reasonCode, final LocalDateTime dateTime) {
+        log.debug("Received moveToCell request for booking id {}, cell location {}, reasonCode {}, date/time {}",
+                bookingId,
+                livingUnitId,
+                reasonCode,
+                dateTime != null ? dateTime.format(ISO_DATE_TIME) : "null");
+
+        validateMoveToCellRequest(reasonCode);
+        final var movementDateTime = dateTime != null ? dateTime : LocalDateTime.now();
```
Comment (matt-crowson-moj) [inline src/main/java/net/syscon/elite/api/resource/impl/OffenderMovementsResourceImpl.java]: I'd suggest using a Clock bean in your service for date/times.  You an then do `LocalDateTime.now(clock)` for example.

You can then take control of time in your tests.  See the `CourtHearingService` as an example.

File: src/main/java/net/syscon/elite/api/resource/OffenderMovementsResource.java (hunk 2 above)
Comment (mikehalmamoj) [reply src/main/java/net/syscon/elite/api/resource/OffenderMovementsResource.java]: Well... the Intellij settings that Peter displayed said to replace 5 or more imports with a *.  
image

File: src/main/java/net/syscon/elite/api/resource/impl/OffenderMovementsResourceImpl.java [hunk 6]
```diff
@@ -33,4 +44,42 @@ public CourtHearing prisonToCourt(final Long bookingId, final Long courtCaseId,
...
+
+        if (reasonCode.equals("123")) {
+            throw new EntityNotFoundException("Simulating a not found for reasonCode '123'");
+        }
+
+        return movementUpdateService.moveToCell(bookingId, livingUnitId, reasonCode, movementDateTime);
+    }
+
+    private void validateMoveToCellRequest(final String reasonCode) {
+        if (StringUtils.isNullOrEmpty(reasonCode)) {
```
Comment (matt-crowson-moj) [inline src/main/java/net/syscon/elite/api/resource/impl/OffenderMovementsResourceImpl.java]: Not sure it is standard (is anything) but Guava Preconditions are useful for runtime exceptions.  Just a slightly lighter touch.

File: src/main/java/net/syscon/elite/api/resource/OffenderMovementsResource.java (hunk 2 above)
Comment (mikehalmamoj) [reply src/main/java/net/syscon/elite/api/resource/OffenderMovementsResource.java]: But if it's GDS policy then agreed - just change the class count in the screenshot above to 999.

**Pull request merged**